docs/analysis.md

```diff
@@ -190,6 +202,171 @@ link to the context in search results that include the target record.
 The `peekRange` property is a range of lines that appears to be
 currently unused.
 
+### Structured Records
+
+Structured records are an attempt to provide richer information about types and
+their relationships.  This is an evolving area of Searchfox, and is subject to
+change.  This expected change also informs the design of the record format,
+which is just a wrapper around an opaque JSON structure as far as `analysis.rs`
+is concerned.  (All other record types are flat with a well known set of
+keys/values.)
+
+We emit structured records at the point of their definition.  Structured
+records reference other structured records by their searchfox symbol identifier.
+A structured record itself won't embed child types (even if they're not visible
+outside the type) but instead reference them (which may involve searchfox
+generating identifiers that have no meaning outside of searchfox, like is done
+for locals).
+
+Because of the realities of compilation, we know a parent class won't
+necessarily know all of its subclasses, so all type references in emitted
+records will generally only be upwards/sideways, never downwards.  We depend on
+cross-referencing for determining sets of children/etc.
+
+So for a class, we might expect the structured record in the analysis file to
+contain:
+- A list of its known super-classes.
+- A list of its fields and members.
+
+But it would not contain:
+- A list of its known sub-classes.  This will be determined by
+  cross-referencing.
+
+#### Bytes and CharUnits
+
+Clang defines a "Character Units" type
+[`CharUnits`](https://clang.llvm.org/doxygen/classclang_1_1CharUnits.html#details)
+that basically means bytes.  Searchfox just calls them bytes and assumes an
+8-bit byte because strings are such a large part of the Firefox codebase that
+calling bytes "char units" just adds terrifying confusion.
+
+#### Formal Hierarchy:
+
+Raw record info.  These are attributes that will be found in the analysis files.
+- `pretty`: The pretty name/identifier for this structured symbol info.
+- `sym`: The searchfox symbol for this symbol.
+- `kind`: A string with one of the following values:
+  - `enum`: TODO: I don't know that this actually is a thing we emit yet?
+  - `class`
+  - `struct`
+  - `union`
+  - `method`: It's a method on a class/struct.  It will have `overrides`.
+  - `function`: A boring function.  No `overrides`.
+  - `field`: A member of a class/struct.  Right now the field record has minimal
+    info with the intent being that the data canonically lives on the parent
+    symbol and this just provides the `parentsym` necessary to get to that info.
+    But this potentially needs more thought.  TODO: Think more on this!
+  - `ipc`: An IPC function where there's a send method and a recv method.  The
```

Comment:
Yeah, I definitely think there are a bunch of semantic relationships that we can capture better and it makes sense for these to not all be one-offs.  And now seems like a great time to have at least a roadmap.

I'm going to plan to ramble on https://bugzilla.mozilla.org/show_bug.cgi?id=1699048 that :valentin filed on prefs/StaticPrefs about this because I think the preferences use case is probably one of the most interesting/complex/far-reaching in mozilla-central.

### IPC specifics

That said, I do want to try and provide some context here for the motivating use-case: 

The key goal for this IPC case was to facilitated the generation of automated class diagrams that express:
- This class lives in the parent/content process
- This class directly interacts with that class directly in the same process
- This class talks to that class via IPC using the PFoo protocol.

From a modeling perspective, for IPC I think the key things are:
- IPC calls are directional
- IPC calls need to be explicitly labeled as having IPC semantics because this is an important detail for users and automated analysis that's concerned about process distinctions, etc.

A definite over-simplification in what's implemented here currently is that there's only a single `targetsym`.  This is largely true since we only have C++ IPC bindings and in most cases there's never 2 concrete implementations of a protocol.  (In fact, I think there's at least one storage protocol where we have 2 concrete implementations and for that reason the IPC bindings have to still use virtual methods.)

Note that one additional semantic addition was the `implKind` where I tried to capture the difference between the WebIDL definition of something and the C++ binding implementation of it, both of which are arguably "defs" but in different ways/at different layers.

Reply:
I ended up rambling at https://bugzilla.mozilla.org/show_bug.cgi?id=1641372#c3 (which is this bug still).